### PR TITLE
Add attack speed attribute to items.xml

### LINKED ISSUE #3395: attackSpeed via items.lua feature.
https://otland.net/threads/c-item-attack-speed-attr-key-not-recognized.262272/

Would be nice to have this on TFS.

Code is old and need a edit.

### PR DESCRIPTION
**Pull request by marmichalski**
### Pull Request Prelude

- [x] I have followed [proper The Forgotten Server code styling][code].
- [x] I have read and understood the [contribution guidelines][cont] before making this PR.
- [x] I am aware that this PR may be closed if the above-mentioned criteria are not fulfilled.

### Changes Proposed

**Issues addressed:** closes #3395.

[cont]: https://github.com/otland/forgottenserver/wiki/Contributing
[code]: https://github.com/otland/forgottenserver/wiki/TFS-Coding-Style-Guide

🤷

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -19,7 +19,7 @@ COPY cmake /usr/src/forgottenserver/cmake/
 COPY src /usr/src/forgottenserver/src/
 COPY CMakeLists.txt /usr/src/forgottenserver/
 WORKDIR /usr/src/forgottenserver/build
-RUN cmake .. && make
+RUN cmake .. && make -j $(($(nproc) / 2 + 1))
```
Comment (Zbizu) [inline Dockerfile:22]: is that intended or pushed by accident?

Comment (marmichalski) [reply Dockerfile:22]: Intended, pushed as separate commit: https://github.com/otland/forgottenserver/pull/3584/commits

**Comment by marmichalski** [comment]
I've included basic attributes (`/attr <name>,<value>`) talkaction with this PR for easier testing. I can move it to it's own PR, if you prefer, as it's not connected in any way with this PR, it's just here for convenience.

File: src/luascript.cpp [hunk 2]
```diff
@@ -2033,6 +2037,7 @@ void LuaScriptInterface::registerFunctions()
 	registerMethod("Game", "getWorldType", LuaScriptInterface::luaGameGetWorldType);
 	registerMethod("Game", "setWorldType", LuaScriptInterface::luaGameSetWorldType);
 
+	registerMethod("Game", "getItemAttributeByName", LuaScriptInterface::luaGameGetItemAttributeByName);
```
Comment (marmichalski) [inline src/luascript.cpp:2040]: no idea where functions from `tools.cpp` should go. `getReturnMessage` is in `Game`, so I've gone with it for `stringToItemAttribute` too.

**Comment by nekiro** [comment]
One pull request one change, so if you add attack speed remove all other code you pushed.
It looks okay, but the lowest you can go is creature think interval, which is 1000 ms.

**Comment by marmichalski** [comment]
> One pull request one change, so if you add attack speed remove all other code you pushed.
> It looks okay, but the lowest you can go is creature think interval, which is 1000 ms.

Not really true (about the interval)

**Comment by nekiro** [comment]
> > One pull request one change, so if you add attack speed remove all other code you pushed.
> > It looks okay, but the lowest you can go is creature think interval, which is 1000 ms.
> 
> Not really true (about the interval). I'll split the changes when I'm home, no problem.

Not true why? doAttacking method executes every 1000 ms, cause its called from checkCreatures loop.

File: Dockerfile (hunk 1 above)
Comment (ranisalt) [reply Dockerfile:22]: I'd rather pass `MAKEFLAGS` as an env/arg and not try to anything smart _inside_ the container.

**Comment by marmichalski** [comment]
> > > One pull request one change, so if you add attack speed remove all other code you pushed.
> > > It looks okay, but the lowest you can go is creature think interval, which is 1000 ms.
> > 
> > 
> > Not really true (about the interval). I'll split the changes when I'm home, no problem.
> 
> Not true why? doAttacking method executes every 1000 ms, cause its called from checkCreatures loop, which loops every 1000 ms.


https://user-images.githubusercontent.com/57528542/130335292-ef454fa8-56dd-41d3-99c9-917be33d043f.mov

Anyway, I do agree that a sane minimum could be set - but not sure what should it be: 250? 100?

**Comment by nekiro** [comment]
> > > > One pull request one change, so if you add attack speed remove all other code you pushed.
> > > > It looks okay, but the lowest you can go is creature think interval, which is 1000 ms.
> > > 
> > > 
> > > Not really true (about the interval). I'll split the changes when I'm home, no problem.
> > 
> > 
> > Not true why? doAttacking method executes every 1000 ms, cause its called from checkCreatures loop, which loops every 1000 ms.
> 
>  Screen.Recording.2021-08-21.at.23.31.38.mov 
> Anyway, I do agree that a sane minimum could be set - but not sure what should it be: 250? 100?

Wait what, you want to lower onthink interval? That's a big nono.

**Comment by marmichalski** [comment]
> > > > > One pull request one change, so if you add attack speed remove all other code you pushed.
> > > > > It looks okay, but the lowest you can go is creature think interval, which is 1000 ms.
> > > > 
> > > > 
> > > > Not really true (about the interval). I'll split the changes when I'm home, no problem.
> > > 
> > > 
> > > Not true why? doAttacking method executes every 1000 ms, cause its called from checkCreatures loop, which loops every 1000 ms.
> > 
> > 
> > Screen.Recording.2021-08-21.at.23.31.38.mov
> > Anyway, I do agree that a sane minimum could be set - but not sure what should it be: 250? 100?
> 
> Wait what, you want to lower onthink interval? That's a big nono.

No. The attack speed that can be set.

**Comment by EPuncker** [comment]
> but not sure what should it be: 250? 100?

200 should be enough

**Review by EPuncker — DISMISSED** [review]

File: src/luascript.cpp [hunk 3]
```diff
@@ -1510,6 +1510,10 @@ void LuaScriptInterface::registerFunctions()
 	registerEnum(ITEM_ATTRIBUTE_CHARGES)
 	registerEnum(ITEM_ATTRIBUTE_FLUIDTYPE)
 	registerEnum(ITEM_ATTRIBUTE_DOORID)
+	registerEnum(ITEM_ATTRIBUTE_DECAYTO)
```
Comment (EPuncker) [inline src/luascript.cpp:1513]: doesn't these belong to the attr PR? 😄

Comment (marmichalski) [reply src/luascript.cpp:1513]: Nope, not really. Those are just missing enums, they are not used in `/attr`.

**Comment by EPuncker** [comment]
Okay, I've tested this today and everything is working just fine, I just have two things to point out:

1. cap the minimum value to 100 (silently?)
2. display the attackspeed attribute at item description? (ASPD:? AS:? Atk Speed:? idk)

**Comment by marmichalski** [comment]
> Okay, I've tested this today and everything is working just fine, I just have two things to point out:
> 
> 1. cap the minimum value to 100 (silently?)
> 2. display the attackspeed attribute at item description? (ASPD:? AS:? Atk Speed:? idk)

1.
```
[Warning - Items::parseItemNode] AttackSpeed lower than 100 for item: 2400
```

2.
```
You see a crossbow (Range:5, Atk Spd:0.1s). //100
You see a club (Atk:7, Def:7, Atk Spd:1s). //1000
You see a club (Atk:7, Def:7, Atk Spd:5.5s).//5500
You see a club (Atk:7, Def:7, Atk Spd:101.521s).// 101521
```

Could `::setprecision` before outputting it, but I'd say... it doesn't matter, whose gonna set attack speed to some random number 🤷

**Review by DSpeichert — APPROVED** [review]

**Pull request merged**